File: code-executor/lib/mcp_client.py
"""
MCP Client Library for Python

Local implementation of MCP protocol client for calling MCP tools from code.
This allows subagents to write code that composes multiple MCP tool calls
without loading all tool schemas into the main context.

Usage:
    from lib.mcp_client import call_mcp_tool

    result = await call_mcp_tool('mcp__filesystem__readFile', {
        'path': '/tmp/data.json'
    })
"""

import asyncio
import json
import os
import subprocess
import sys
from pathlib import Path
from typing import Dict, Any, List, Tuple
from uuid import uuid4


class MCPError(Exception):
    """MCP tool call error"""
    pass


def parse_mcp_tool_name(tool_name: str) -> Tuple[str, str]:
    """
    Parse MCP tool name into server and tool components.

    Args:
        tool_name: Full MCP tool name (e.g., 'mcp__filesystem__readFile')

    Returns:
        Tuple of (server_name, tool_name)

    Raises:
        ValueError: If tool name format is invalid
    """
    parts = tool_name.split('__')

    if len(parts) != 3 or parts[0] != 'mcp':
        raise ValueError(
            f"Invalid MCP tool name format: {tool_name}. "
            f"Expected: mcp__<server>__<tool>"
        )

    return parts[1], parts[2]


async def get_mcp_server_config(server_name: str) -> Dict[str, Any]:
    """
    Get MCP server configuration from environment or config file.

    Args:
        server_name: Name of the MCP server

    Returns:
        Server configuration dict

    Raises:
        FileNotFoundError: If config file not found
        MCPError: If server not configured
    """
    # Try environment variable first
    config_path = os.getenv('MCP_CONFIG_PATH')
    if not config_path:
        config_path = os.path.join(os.path.expanduser('~'), '.mcp.json')

    config_path = Path(config_path)

    if not config_path.exists():
        raise FileNotFoundError(
            f"MCP config file not found: {config_path}"
        )

    try:
        with open(config_path, 'r') as f:
            config = json.load(f)
    except json.JSONDecodeError as e:
        raise MCPError(f"Invalid JSON in MCP config: {e}")

    if 'mcpServers' not in config:
        raise MCPError(f"No 'mcpServers' key in config: {config_path}")

    if server_name not in config['mcpServers']:
        raise MCPError(
            f"MCP server '{server_name}' not found in config: {config_path}"
        )

    return config['mcpServers'][server_name]


async def call_mcp_tool_via_stdio(
    server_config: Dict[str, Any],
    tool_name: str,
    parameters: Dict[str, Any]
) -> Any:
    """
    Call an MCP tool via stdio protocol.

    Args:
        server_config: Server configuration dict
        tool_name: Tool name (without mcp__ prefix)
        parameters: Tool parameters

    Returns:
        Tool result

    Raises:
        MCPError: If tool call fails
    """
    command = server_config['command']
    args = server_config.get('args', [])

    # Create MCP request
    request = {
        'jsonrpc': '2.0',
        'id': str(uuid4()),
        'method': 'tools/call',
        'params': {
            'name': tool_name,
            'arguments': parameters
        }
    }

    request_json = json.dumps(request) + '\n'

    # Start MCP server process
    try:
        process = await asyncio.create_subprocess_exec(
            command,
            *args,
            stdin=asyncio.subprocess.PIPE,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE
        )
    except FileNotFoundError:
        raise MCPError(f"MCP server command not found: {command}")

    # Send request
    stdout, stderr = await process.communicate(request_json.encode('utf-8'))

    if stderr:
        stderr_text = stderr.decode('utf-8', errors='replace')
        print(f"MCP server stderr: {stderr_text}", file=sys.stderr)

    # Parse response
    stdout_text = stdout.decode('utf-8', errors='replace')
    lines = stdout_text.strip().split('\n')

    for line in lines:
        if not line.strip():
            continue

        try:
            response = json.loads(line)

            if 'error' in response:
                raise MCPError(
                    f"MCP tool error: {json.dumps(response['error'])}"
                )

            if 'result' in response:
                return response['result']
        except json.JSONDecodeError:
            # Skip non-JSON lines (might be server logs)
            continue

    raise MCPError('No valid MCP response received')


async def call_mcp_tool(
    tool_name: str,
    parameters: Dict[str, Any] = None
) -> Any:
    """
    Call an MCP tool.

    Args:
        tool_name: Full MCP tool name (e.g., 'mcp__filesystem__readFile')
        parameters: Tool parameters as dict

    Returns:
        Tool result

    Example:
        >>> content = await call_mcp_tool('mcp__filesystem__readFile', {
        ...     'path': '/tmp/data.json'
        ... })
    """
    if parameters is None:
        parameters = {}

    server_name, tool = parse_mcp_tool_name(tool_name)

    # Get server configuration
    server_config = await get_mcp_server_config(server_name)

    # Call tool via stdio (most common MCP transport)
    # Future: Add support for HTTP transport
    try:
        result = await call_mcp_tool_via_stdio(
            server_config,
            tool,
            parameters
        )
        return result
    except Exception as e:
        raise MCPError(f"Failed to call MCP tool {tool_name}: {e}") from e


async def call_mcp_tools_parallel(
    calls: List[Dict[str, Any]]
) -> List[Any]:
    """
    Call multiple MCP tools in parallel.

    Args:
        calls: List of dicts with 'tool' and 'parameters' keys

    Returns:
        List of results in same order

    Example:
        >>> results = await call_mcp_tools_parallel([
        ...     {'tool': 'mcp__filesystem__readFile', 'parameters': {'path': '/tmp/a.json'}},
        ...     {'tool': 'mcp__filesystem__readFile', 'parameters': {'path': '/tmp/b.json'}}
        ... ])
    """
    tasks = [
        call_mcp_tool(call['tool'], call.get('parameters', {}))
        for call in calls
    ]
    return await asyncio.gather(*tasks)


async def call_mcp_tools_parallel_safe(
    calls: List[Dict[str, Any]]
) -> List[Dict[str, Any]]:
    """
    Call multiple MCP tools in parallel with graceful error handling.

    Args:
        calls: List of dicts with 'tool' and 'parameters' keys

    Returns:
        List of result dicts with 'success', 'data'/'error' keys

    Example:
        >>> results = await call_mcp_tools_parallel_safe([
        ...     {'tool': 'mcp__api__fetch', 'parameters': {'url': '...'}},
        ...     {'tool': 'mcp__api__fetch', 'parameters': {'url': '...'}}
        ... ])
        >>> for i, result in enumerate(results):
        ...     if result['success']:
        ...         print(f"Success: {result['data']}")
        ...     else:
        ...         print(f"Failed: {result['error']}")
    """
    tasks = [
        call_mcp_tool(call['tool'], call.get('parameters', {}))
        for call in calls
    ]

    results = await asyncio.gather(*tasks, return_exceptions=True)

    return [
        {'success': True, 'data': result}
        if not isinstance(result, Exception)
        else {'success': False, 'error': str(result)}
        for result in results
    ]

File: code-executor/lib/mcp-client.ts
/**
 * MCP Client Library for TypeScript/Deno
 *
 * Local implementation of MCP protocol client for calling MCP tools from code.
 * This allows subagents to write code that composes multiple MCP tool calls
 * without loading all tool schemas into the main context.
 *
 * Usage:
 *   import { callMCPTool } from './lib/mcp-client.ts';
 *
 *   const result = await callMCPTool('mcp__filesystem__readFile', {
 *     path: '/tmp/data.json'
 *   });
 */

interface MCPToolCall {
  tool: string;
  parameters: Record<string, unknown>;
}

interface MCPToolResult {
  success: boolean;
  data?: unknown;
  error?: string;
}

interface MCPServerConfig {
  command: string;
  args?: string[];
}

/**
 * Parse MCP tool name into server and tool components
 */
function parseMCPToolName(toolName: string): { server: string; tool: string } {
  const parts = toolName.split("__");

  if (parts.length !== 3 || parts[0] !== "mcp") {
    throw new Error(
      `Invalid MCP tool name format: ${toolName}. Expected: mcp__<server>__<tool>`,
    );
  }

  return {
    server: parts[1],
    tool: parts[2],
  };
}

/**
 * Get MCP server configuration from environment or config file
 */
async function getMCPServerConfig(
  serverName: string,
): Promise<MCPServerConfig> {
  // Try environment variable first
  const configPath = Deno.env.get("MCP_CONFIG_PATH") ||
    Deno.env.get("HOME") + "/.mcp.json";

  try {
    const configContent = await Deno.readTextFile(configPath);
    const config = JSON.parse(configContent);

    if (!config.mcpServers || !config.mcpServers[serverName]) {
      throw new Error(
        `MCP server '${serverName}' not found in config: ${configPath}`,
      );
    }

    return config.mcpServers[serverName];
  } catch (error) {
    const err = error as Error;
    throw new Error(
      `Failed to load MCP config from ${configPath}: ${err.message}`,
    );
  }
}

/**
 * Call an MCP tool via stdio protocol
 */
async function callMCPToolViaStdio(
  serverConfig: MCPServerConfig,
  toolName: string,
  parameters: Record<string, unknown>,
): Promise<unknown> {
  const command = new Deno.Command(serverConfig.command, {
    args: serverConfig.args || [],
    stdin: "piped",
    stdout: "piped",
    stderr: "piped",
  });

  const process = command.spawn();
  const writer = process.stdin.getWriter();

  // Send MCP request
  const request = {
    jsonrpc: "2.0",
    id: crypto.randomUUID(),
    method: "tools/call",
    params: {
      name: toolName,
      arguments: parameters,
    },
  };

  const encoder = new TextEncoder();
  await writer.write(encoder.encode(JSON.stringify(request) + "\n"));
  await writer.close();

  // Read response
  const output = await process.output();
  const decoder = new TextDecoder();
  const stdout = decoder.decode(output.stdout);
  const stderr = decoder.decode(output.stderr);

  if (stderr) {
    console.error("MCP server stderr:", stderr);
  }

  // Parse JSON-RPC response
  const lines = stdout.trim().split("\n");
  for (const line of lines) {
    if (!line.trim()) continue;

    try {
      const response = JSON.parse(line);

      if (response.error) {
        throw new Error(`MCP tool error: ${JSON.stringify(response.error)}`);
      }

      if (response.result) {
        return response.result;
      }
    } catch (_parseError) {
      // Skip non-JSON lines (might be server logs)
      continue;
    }
  }

  throw new Error("No valid MCP response received");
}

/**
 * Main function to call an MCP tool
 *
 * @param toolName - Full MCP tool name (e.g., 'mcp__filesystem__readFile')
 * @param parameters - Tool parameters as object
 * @returns Tool result
 *
 * @example
 * ```typescript
 * const content = await callMCPTool('mcp__filesystem__readFile', {
 *   path: '/tmp/data.json'
 * });
 * ```
 */
export async function callMCPTool(
  toolName: string,
  parameters: Record<string, unknown> = {},
): Promise<unknown> {
  const { server, tool } = parseMCPToolName(toolName);

  // Get server configuration
  const serverConfig = await getMCPServerConfig(server);

  // Call tool via stdio (most common MCP transport)
  // Future: Add support for HTTP transport
  try {
    const result = await callMCPToolViaStdio(serverConfig, tool, parameters);
    return result;
  } catch (error) {
    const err = error as Error;
    throw new Error(
      `Failed to call MCP tool ${toolName}: ${err.message}`,
    );
  }
}

/**
 * Call multiple MCP tools in parallel
 *
 * @param calls - Array of tool calls
 * @returns Array of results in same order
 *
 * @example
 * ```typescript
 * const [file1, file2] = await callMCPToolsParallel([
 *   { tool: 'mcp__filesystem__readFile', parameters: { path: '/tmp/a.json' } },
 *   { tool: 'mcp__filesystem__readFile', parameters: { path: '/tmp/b.json' } }
 * ]);
 * ```
 */
export async function callMCPToolsParallel(
  calls: MCPToolCall[],
): Promise<unknown[]> {
  return await Promise.all(
    calls.map(({ tool, parameters }) => callMCPTool(tool, parameters)),
  );
}

/**
 * Call multiple MCP tools in parallel with graceful error handling
 *
 * @param calls - Array of tool calls
 * @returns Array of results (settled promises)
 *
 * @example
 * ```typescript
 * const results = await callMCPToolsParallelSettled([
 *   { tool: 'mcp__api__fetch', parameters: { url: '...' } },
 *   { tool: 'mcp__api__fetch', parameters: { url: '...' } }
 * ]);
 *
 * results.forEach((result, i) => {
 *   if (result.status === 'fulfilled') {
 *     console.log(`Success: ${result.value}`);
 *   } else {
 *     console.error(`Failed: ${result.reason}`);
 *   }
 * });
 * ```
 */
export async function callMCPToolsParallelSettled(
  calls: MCPToolCall[],
): Promise<PromiseSettledResult<unknown>[]> {
  return await Promise.allSettled(
    calls.map(({ tool, parameters }) => callMCPTool(tool, parameters)),
  );
}
